File: python/src/train/models/full_mix/block.py
"""Full-Mix block: parameterized full token mixing + GLU-improved P-FFNs."""

import torch
import torch.nn as nn
import torch.nn.functional as F


class PerTokenGluFfn(nn.Module):
    """Dedicated GLU-style FFN per token."""

    def __init__(self, num_tokens: int, token_dim: int, hidden_factor: float) -> None:
        super().__init__()
        if hidden_factor <= 0:
            raise ValueError("hidden_factor must be > 0")
        hidden_dim = max(1, int(token_dim * hidden_factor))
        self.up = nn.ModuleList(nn.Linear(token_dim, hidden_dim) for _ in range(num_tokens))
        self.gate = nn.ModuleList(nn.Linear(token_dim, hidden_dim) for _ in range(num_tokens))
        self.down = nn.ModuleList(nn.Linear(hidden_dim, token_dim) for _ in range(num_tokens))

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        outputs = []
        for token_idx, (up, gate, down) in enumerate(
            zip(self.up, self.gate, self.down, strict=False)
        ):
            token = x[:, token_idx, :]
            hidden = F.gelu(up(token)) * torch.sigmoid(gate(token))
            outputs.append(down(hidden).unsqueeze(1))
        return torch.cat(outputs, dim=1)


class FullMixBlock(nn.Module):
    """RankElastor-style full token mixing block."""

    def __init__(
        self,
        token_dim: int,
        num_tokens: int,
        hidden_factor: float,
    ) -> None:
        super().__init__()
        if token_dim <= 0:
            raise ValueError("token_dim must be > 0")
        if num_tokens <= 0:
            raise ValueError("num_tokens must be > 0")
        if hidden_factor <= 0:
            raise ValueError("hidden_factor must be > 0")
        self.token_dim = token_dim
        self.num_tokens = num_tokens
        flat_dim = token_dim * num_tokens
        self.full_mixing = nn.Linear(flat_dim, flat_dim)
        self.norm_mixing = nn.LayerNorm(token_dim)
        self.pffn = PerTokenGluFfn(num_tokens, token_dim, hidden_factor)
        self.norm_ffn = nn.LayerNorm(token_dim)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        batch_size = x.shape[0]
        mixed = self.full_mixing(x.reshape(batch_size, -1))
        mixed = mixed.reshape(batch_size, self.num_tokens, self.token_dim)
        s = self.norm_mixing(mixed + x)
        return self.norm_ffn(self.pffn(s) + s)

File: src/models/full_mix/block.rs
//! Full-Mix block: parameterized full token mixing + GLU-improved per-token FFNs.

use candle_core::{Result, Tensor};
use candle_nn::{layer_norm, linear, LayerNorm, Linear, Module, VarBuilder};

/// Dedicated GLU-style FFN per token (paper Eq. 4).
///
///   output_t = (GELU(M_t · W₁) ⊙ (M_t · W₂)) · W₃ + M_t · Wᵣ
///
/// where Wᵣ is a learnable residual projection (not merely an identity shortcut).
pub struct PerTokenGluFfn {
    up: Vec<Linear>,
    gate: Vec<Linear>,
    down: Vec<Linear>,
    skip: Vec<Linear>,
}

impl PerTokenGluFfn {
    /// Construct per-token GLU FFN modules.
    pub fn new(
        num_tokens: usize,
        token_dim: usize,
        hidden_factor: f64,
        vb: VarBuilder,
    ) -> Result<Self> {
        if hidden_factor <= 0.0 {
            candle_core::bail!("hidden_factor must be > 0");
        }
        let hidden_dim = ((token_dim as f64) * hidden_factor).floor().max(1.0) as usize;
        let mut up = Vec::with_capacity(num_tokens);
        let mut gate = Vec::with_capacity(num_tokens);
        let mut down = Vec::with_capacity(num_tokens);
        let mut skip = Vec::with_capacity(num_tokens);
        for token_idx in 0..num_tokens {
            up.push(linear(
                token_dim,
                hidden_dim,
                vb.pp("up").pp(token_idx.to_string()),
            )?);
            gate.push(linear(
                token_dim,
                hidden_dim,
                vb.pp("gate").pp(token_idx.to_string()),
            )?);
            down.push(linear(
                hidden_dim,
                token_dim,
                vb.pp("down").pp(token_idx.to_string()),
            )?);
            skip.push(linear(
                token_dim,
                token_dim,
                vb.pp("skip").pp(token_idx.to_string()),
            )?);
        }
        Ok(Self {
            up,
            gate,
            down,
            skip,
        })
    }

    /// Forward over `[batch, num_tokens, token_dim]`.
    ///
    /// Paper Eq. 4: Z_t = (GELU(M_t·W₁) ⊙ (M_t·W₂))·W₃ + M_t·Wᵣ
    /// Gate path uses identity activation (no sigmoid).
    pub fn forward(&self, x: &Tensor) -> Result<Tensor> {
        let mut outputs = Vec::with_capacity(self.up.len());
        for token_idx in 0..self.up.len() {
            let token = x.narrow(1, token_idx, 1)?.squeeze(1)?;
            let up = self.up[token_idx].forward(&token)?.gelu()?;
            let gate = self.gate[token_idx].forward(&token)?; // identity activation
            let hidden = up.mul(&gate)?;
            let gated = self.down[token_idx].forward(&hidden)?;
            let residual = self.skip[token_idx].forward(&token)?;
            outputs.push(gated.add(&residual)?.unsqueeze(1)?);
        }
        Tensor::cat(&outputs, 1)
    }
}

/// Full-Mix block with learnable mixing over flattened token representations.
pub struct FullMixBlock {
    token_dim: usize,
    num_tokens: usize,
    full_mixing: Linear,
    norm_mixing: LayerNorm,
    pffn: PerTokenGluFfn,
    norm_ffn: LayerNorm,
}

impl FullMixBlock {
    /// Construct a Full-Mix block.
    pub fn new(
        token_dim: usize,
        num_tokens: usize,
        hidden_factor: f64,
        vb: VarBuilder,
    ) -> Result<Self> {
        if token_dim == 0 {
            candle_core::bail!("token_dim must be > 0");
        }
        if num_tokens == 0 {
            candle_core::bail!("num_tokens must be > 0");
        }
        if hidden_factor <= 0.0 {
            candle_core::bail!("hidden_factor must be > 0");
        }
        let flat_dim = token_dim * num_tokens;
        Ok(Self {
            token_dim,
            num_tokens,
            full_mixing: linear(flat_dim, flat_dim, vb.pp("full_mixing"))?,
            norm_mixing: layer_norm(token_dim, 1e-5, vb.pp("norm_mixing"))?,
            pffn: PerTokenGluFfn::new(num_tokens, token_dim, hidden_factor, vb.pp("pffn"))?,
            norm_ffn: layer_norm(token_dim, 1e-5, vb.pp("norm_ffn"))?,
        })
    }

    /// Forward over `[batch, num_tokens, token_dim]`.
    pub fn forward(&self, x: &Tensor) -> Result<Tensor> {
        let batch_size = x.dim(0)?;
        let flat_dim = self.num_tokens * self.token_dim;
        let mixed = self
            .full_mixing
            .forward(&x.reshape((batch_size, flat_dim))?)?
            .reshape((batch_size, self.num_tokens, self.token_dim))?;
        let s = self.norm_mixing.forward(&mixed.add(x)?)?;
        self.norm_ffn.forward(&self.pffn.forward(&s)?.add(&s)?)
    }
}
